bindings/aws/s3/s3.go

```diff
@@ -44,39 +73,237 @@ func (s *AWSS3) Init(metadata bindings.Metadata) error {
+	count, counterr := strconv.ParseInt(req.Metadata[metadataKeyCount],10,64)
+
+	if(offseterr == nil && counterr == nil) {
+		s3range := "bytes=" + req.Metadata[metadataKeyOffset] + "-" + strconv.FormatInt(offset + count -1, 10)
+
+		requestInput := s3.GetObjectInput{
+			Bucket: aws.String(s.metadata.Bucket),
+			Key:    aws.String(req.Metadata[metadataKeyBlobName]),
+			Range:  aws.String(s3range),
+		}
+
+		s.downloader.Download(b,&requestInput)
+	} else{
+
+		requestInput := s3.GetObjectInput{
+			Bucket: aws.String(s.metadata.Bucket),
+			Key:    aws.String(req.Metadata[metadataKeyBlobName]),
+		}
+
+		s.downloader.Download(b,&requestInput)
+	}
+
+	return &bindings.InvokeResponse{
+		Data:     b.Bytes(),
+		Metadata: metadata,
+	}, nil
+}
+
+func (s *AWSS3) head(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	requestInput := s3.HeadObjectInput{
+		Bucket: aws.String(s.metadata.Bucket),
+		Key:    aws.String(req.Metadata[metadataKeyBlobName]),
+	}
+
+
+	result, err := s.S3.HeadObject(&requestInput)
+
+	metadata := make(map[string]string)
+	metadata["Content-Length"] = strconv.FormatInt(aws.Int64Value(result.ContentLength), 10)
+
+	return &bindings.InvokeResponse{
+	Data:     nil,
+	Metadata: metadata,
+	}, err
+}
+
+func (s *AWSS3) initbackup(key string ) (error){
+	backup, ok := s.backups[key]
+
+	if(ok) {
+		return nil
+	}
+
+	MultipartResponse, err := s.S3.CreateMultipartUpload(&s3.CreateMultipartUploadInput{
+		Bucket: aws.String(s.metadata.Bucket),
+		Key:    aws.String(key),
 	})
+	if(err != nil){
+		return err
+	}
+	backup.MultipartResponse = MultipartResponse
+	backup.PartNum = 1
+	backup.CompletedParts = nil
+	s.backups[key] = backup
+
+	return nil
+}
+
+func (s *AWSS3) put(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	buffer, _ := hex.DecodeString(req.Metadata[metadataKeyData])
+	key, _ := req.Metadata[metadataKeyBlobName]
+	err := s.initbackup(key)
+
+	if(err != nil){
+		return nil,err
+	}
+
+	backup := s.backups[key]
+	uploadResp, err := s.S3.UploadPart(&s3.UploadPartInput{
+		Body:          bytes.NewReader(buffer),
+		Bucket:        backup.MultipartResponse.Bucket,
+		Key:           backup.MultipartResponse.Key,
+		PartNumber:    aws.Int64(int64(s.backups[key].PartNum)),
+		UploadId:      backup.MultipartResponse.UploadId,
+		ContentLength: aws.Int64(int64(len(buffer))),
+	})
+
+	if err != nil {
+		s.S3.AbortMultipartUpload(&s3.AbortMultipartUploadInput{
+			Bucket:   backup.MultipartResponse.Bucket,
+			Key:      backup.MultipartResponse.Key,
+			UploadId: backup.MultipartResponse.UploadId,
+		})
+		delete(s.backups,key)
+	} else{
+		CompletedPart := &s3.CompletedPart{
+			ETag:       uploadResp.ETag,
+			PartNumber: aws.Int64(int64(backup.PartNum)),
+		}
+		Offset, _ := strconv.ParseInt(req.Metadata[metadataKeyOffset],10,64)
+		backup.CompletedParts = append(backup.CompletedParts, Part{CompletedPart: CompletedPart, Offset: Offset})
+		backup.PartNum++
```

Comment:
if multiple requests come in the same time, is there going to be a race condition?

Reply:
Just to update with offline conclusion: No need to increment this, it will be passed by the client.